drivers/bus/acpi/compbatt/compbatt.c

```diff
@@ -202,40 +508,635 @@ CompBattDisableStatusNotify(IN PCOMPBATT_DEVICE_EXTENSION DeviceExtension)
+    if (!NT_SUCCESS(Status))
+    {
+        if (CompBattDebug & COMPBATT_DEBUG_ERR)
+            DbgPrint("CompBatt: Failed to refresh composite battery's status (Status 0x%08lx)\n", Status);
+
+        return Status;
+    }
+
+    /* Print out battery status data that has been polled */
+    if (CompBattDebug & COMPBATT_DEBUG_INFO)
+        DbgPrint("CompBatt: Latest composite battery status (when setting notify status)\n"
+                 "          PowerState -> 0x%lx\n"
+                 "          Capacity -> %u\n"
+                 "          Voltage -> %u\n"
+                 "          Rate -> %d\n",
+                 BatteryStatus.PowerState,
+                 BatteryStatus.Capacity,
+                 BatteryStatus.Voltage,
+                 BatteryStatus.Rate);
+
+    /* Calculate the high and low capacity differences based on the real summed capacity of the composite */
+    LowCapDifference = DeviceExtension->BatteryStatus.Capacity - BatteryNotify->LowCapacity;
+    HighCapDifference = BatteryNotify->HighCapacity - DeviceExtension->BatteryStatus.Capacity;
+
+    /* Cache the notification parameters provided for later usage when polling for battery status */
+    DeviceExtension->WaitNotifyStatus.PowerState = BatteryNotify->PowerState;
+    DeviceExtension->WaitNotifyStatus.LowCapacity = BatteryNotify->LowCapacity;
+    DeviceExtension->WaitNotifyStatus.HighCapacity = BatteryNotify->HighCapacity;
+
+    /* Toggle the valid notify flag as these are the newer notification settings */
+    DeviceExtension->Flags |= COMPBATT_STATUS_NOTIFY_SET;
+
+    /*
+     * Get the number of currently linked batteries to composite and total rate,
+     * we will use these counters later to determine the wait values for each
+     * individual battery.
+     */
+    BadBattery = CompBattCalculateTotalRateAndLinkedBatteries(DeviceExtension,
+                                                              &TotalRate,
+                                                              &BatteriesCount);
+
+    /*
+     * Of course we have to be sure that we have at least one battery linked
+     * with the composite battery at this time of getting invoked to set new
+     * notification wait settings.
+     */
+    ASSERT(BatteriesCount != 0);
+
+    /* Walk over the linked batteries list and set up new wait configuration settings */
+    ExAcquireFastMutex(&DeviceExtension->Lock);
+    ListHead = &DeviceExtension->BatteryList;
+    for (NextEntry = ListHead->Flink;
+         NextEntry != ListHead;
+         NextEntry = NextEntry->Flink)
+    {
+        /* Acquire the remove lock so this battery does not disappear under us */
+        BatteryData = CONTAINING_RECORD(NextEntry, COMPBATT_BATTERY_DATA, BatteryLink);
+        if (!NT_SUCCESS(IoAcquireRemoveLock(&BatteryData->RemoveLock, BatteryData->Irp)))
+            continue;
+
+        /* Now release the device lock since the battery can't go away */
+        ExReleaseFastMutex(&DeviceExtension->Lock);
```

Comment:
Question (here and the other similar loops below):
What's the reason for temporarily releasing the mutex? Is it because you cancel the Irp (IoCancelIrp call) below, or (for the other loops), because you're doing an IOCTL to the lower battery drivers?

Reply:
Apart that an individual battery cannot go away as the I/O remove lock is held so the Composite Battery mutex is no longer needed, it's also for the reasons you stated. 

Cancelling IRPs is done with a raised IRQL to dispatch, of which at this point your code of the current calling thread no longer runs at APC_LEVEL, which it's the main rule for fast mutexes. Also you wouldn't want to hold the mutex for the calling thread as it issues IOCTLs, which means you're going to call the lower device drivers in the stack.